controllers/crdmigrator/crd_migrator.go

```diff
@@ -0,0 +1,549 @@
+	migrationConfig, ok := r.configByCRDName[req.Name]
+	if !ok {
+		// If there is no migrationConfig for this CRD, do nothing.
+		return ctrl.Result{}, nil
+	}
+
+	crdPartial := &metav1.PartialObjectMetadata{}
+	crdPartial.SetGroupVersionKind(apiextensionsv1.SchemeGroupVersion.WithKind("CustomResourceDefinition"))
+	if err := r.Client.Get(ctx, req.NamespacedName, crdPartial); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	currentGeneration := strconv.FormatInt(crdPartial.GetGeneration(), 10)
+	if observedGeneration, ok := crdPartial.Annotations[clusterv1.CRDMigrationObservedGenerationAnnotation]; ok &&
+		currentGeneration == observedGeneration {
+		// If the current generation was already reconciled/observed, do nothing.
+		return ctrl.Result{}, nil
+	}
+
+	crd := &apiextensionsv1.CustomResourceDefinition{}
+	if err := r.APIReader.Get(ctx, req.NamespacedName, crd); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	// Note: The CRD might have a newer generation than PartialObjectMeta above, we should use the
+	// generation of the CRD we actually use for reconciliation.
+	currentGeneration = strconv.FormatInt(crd.GetGeneration(), 10)
+
+	storageVersion, err := storageVersionForCRD(crd)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	originalCRD := crd.DeepCopy()
+	defer func() {
+		if reterr == nil {
+			if crd.Annotations == nil {
+				crd.Annotations = map[string]string{}
+			}
+			crd.Annotations[clusterv1.CRDMigrationObservedGenerationAnnotation] = currentGeneration
+			if err := r.Client.Patch(ctx, crd, client.MergeFrom(originalCRD)); err != nil {
+				reterr = kerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch CustomResourceDefinition %s", crd.Name)})
+			}
+		}
+	}()
+
+	var customResourceObjects []client.Object
+	if r.crdMigrationPhasesToRun.Has(StorageVersionMigrationPhase) && storageVersionMigrationRequired(crd, storageVersion) ||
+		r.crdMigrationPhasesToRun.Has(CleanupManagedFieldsPhase) {
+		// Get CustomResources only if we actually are going to run one of the phases.
+		// Note: This relies on that the version that is storageVersion is also served.
+		var err error
+		customResourceObjects, err = r.listCustomResources(ctx, crd, migrationConfig, storageVersion)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// If phase should be run and .status.storedVersions != [storageVersion], run storage version migration.
+	if r.crdMigrationPhasesToRun.Has(StorageVersionMigrationPhase) && storageVersionMigrationRequired(crd, storageVersion) {
+		if err := r.reconcileStorageVersionMigration(ctx, crd, customResourceObjects, storageVersion); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		crd.Status.StoredVersions = []string{storageVersion}
+		if err := r.Client.Status().Patch(ctx, crd, client.MergeFrom(originalCRD)); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to patch CustomResourceDefinition %s", crd.Name)
+		}
```

Comment:
optional: after we know that we successfully migrated all objects for the crd, we could drop them from the cache already. But I guess the cache entries are too small to be worth it removing it from cache before hitting the 10m ttl.

But maybe an option to think about if we see the ramp-up phase of a controlelr to be too hard.

Maybe even worth to measure in some way how long it mostly takes for the migration to finish and just reduce the ttl of 10m, because after we set the annotation on the CRD, its very unlikely to have the case where we need the cached objects.

Reply:
Hm yup. I would avoid optimizing this for now. Should be only a few kb I think and only for a short period of time

> Maybe even worth to measure in some way how long it mostly takes for the migration to finish and just reduce the ttl of 10m, because after we set the annotation on the CRD, its very unlikely to have the case where we need the cached objects.

I took a closer look and we're at roughly 50-100 objects per second. So 4k Machines => 40s-80s

I would keep the current 10m and probably rather increase then reduce. If we actually hit errors and exponential backoff we would need more.